presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java

```diff
@@ -0,0 +1,175 @@
+    }
+
+    @CombineFunction
+    public static void combine(
+            @AggregationState PrecisionRecallState state,
+            @AggregationState PrecisionRecallState otherState)
+    {
+        state.setNull(state.getNull() || otherState.getNull());
+        if (state.getNull()) {
+            return;
+        }
+
+        if (state.getTrueWeights() == null && otherState.getTrueWeights() != null) {
+            state.setTrueWeights(otherState.getTrueWeights().copy());
+            state.setFalseWeights(otherState.getFalseWeights().copy());
+            return;
+        }
+        if (state.getTrueWeights() != null && otherState.getTrueWeights() != null) {
+            state.getTrueWeights().mergeWith(otherState.getTrueWeights());
+            state.getFalseWeights().mergeWith(otherState.getFalseWeights());
+        }
+    }
+
+    @OutputFunction("map(double,double)")
+    public static void output(@AggregationState PrecisionRecallState state, BlockBuilder out)
+    {
+        if (state.getNull()) {
+            out.appendNull();
+            return;
+        }
+
+        SortedMap<Double, Double> result;
+        if (state.getTrueWeights() == null) {
+            result = PrecisionRecallAggregation.getEmptyResultMap();
+        }
+        else {
+            result = PrecisionRecallAggregation.getResultMap(state);
+        }
+
+        BlockBuilder entryBuilder = out.beginBlockEntry();
+        Iterator<Map.Entry<Double, Double>> resultIt = result.entrySet().iterator();
+        while (resultIt.hasNext()) {
+            final Map.Entry<Double, Double> entry = resultIt.next();
+            final Double recall = entry.getKey();
+            final Double precision = entry.getValue();
+            DoubleType.DOUBLE.writeDouble(entryBuilder, recall);
+            DoubleType.DOUBLE.writeDouble(entryBuilder, precision);
+        }
+        out.closeEntry();
+    }
+
+    private PrecisionRecallAggregation() {}
+
+    private static Double safeDiv(Double nom, Double denom)
+    {
+        return denom == 0.0 ? 1.0 : nom / denom;
+    }
+
+    private static SortedMap<Double, Double> getResultMap(@AggregationState PrecisionRecallState state)
+    {
+        final Double totalTrueWeight = state.getTrueWeights().getBuckets().values().stream()
+                .mapToDouble(Double::doubleValue)
+                .sum();
+        if (totalTrueWeight == 0.0) {
+            return PrecisionRecallAggregation.getEmptyResultMap();
+        }
+
+        final Double totalFalseWeight = state.getFalseWeights().getBuckets().values().stream()
+                .mapToDouble(Double::doubleValue)
+                .sum();
+
+        SortedMap<Double, Double> result = new TreeMap<>();
+        Double runningWeight = 0.0;
+        Double runningTrueWeight = 0.0;
+        Iterator<Map.Entry<Double, Double>> trueIt = state.getTrueWeights().getBuckets().entrySet().iterator();
```

Comment:
The suggestion on using an iterator rather than a map may also make this part a bit more clear, since the key and values have specific meanings which you can't know just by looking at the generics.